Add flag to administratively enable APIs on bootstrap

joestringer opened the pull request:
Depends on https://github.com/cilium/cilium/pull/24967 .

Add a new (hidden) flag, `enable-cilium-api-server-access`, which accepts a list of APIs
that the daemon should allow REST API clients to mutate.

This flag can be useful to restrict the sorts of changes that can be
made locally directly from the node or from users with access to the
Cilium container. Many of these have been added over time to assist
development and debugging purposes, and may not be necessary in
production environments.

Set the default to allow "all" API modifications (`*`) so that there is no
change in behaviour for existing deployments.

In order to successfully use this flag, it is likely that at least the following options must be configured in order to allow cilium-cni to provision pods, and to ensure that the health of the agent can be assessed by Kubernetes:

- `GetConfig`
- `GetHealthz`
- `PutEndpointID`
- `DeleteEndpointID`
- `PostIPAM`
- `DeleteIPAMIP`

Users may additionally see benefit in providing `Get*` as a wildcard in order to allow any read-only APIs to be accessed, which can be very useful for debugging. Occasionally, additional flush APIs can also be very useful in order to mitigate issues at runtime, so this option should not be used unless you have a pretty strong understanding of the APIs you are likely to rely upon.

For completeness, similar flags are also added for the cilium-health and operator APIs, though the usefulness there is more limited.

joestringer committed:
api: Add helper to administratively disable APIs

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer commented on contrib/scripts/cilium-api-check.sh:
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+# Find all files with API handlers for non-GET APIs
+API_HANDLERS=( $(grep -Rl "Handle[^) ]* [^G][^.]*Param" daemon/cmd/*.go) )
```
The path here might end up breaking in future if we refactor the API handlers (cc @joamaki). I just noticed this also doesn't catch the cilium-health API which also has a PUT, since that is under `pkg/health/server/status.go`.

Maybe we can change the filepath filter here, I'm trying to avoid `api/`, `Documentation/` and `vendor/` directories and tags in the root of the repository.

joestringer replied:
^ Fixing in the newest version with `grep --exclude-dir` :sparkles:

tklauser commented on Makefile:
```diff
@@ -647,7 +648,11 @@ install-manpages: ## Install manpages the Cilium CLI.
 	cp man/* /usr/local/share/man/man1/
 	mandb
 
+api-check:
```
Should this have a doc string to show up in `make help`?

```suggestion
api-check: ## Check generated API for mutation permissions checks
```

tklauser commented on contrib/scripts/cilium-api-gen.sh:
```diff
@@ -0,0 +1,159 @@
+        debug "Generating code for $api into $target"
+        preamble "${packages[@]}" > $target
+        flags "${structures[@]}" >> $target
+        defaults "${structures[@]}" >> $target
```
I wonder whether we should run `gofmt $target` just to be on the safe side.

tklauser reviewed (approved):
LGTM with two non-blocking suggestions. Feel free to apply or ignore as you see fit.

joamaki commented on daemon/cmd/endpoint.go:
```diff
@@ -533,14 +534,18 @@ func (h *putEndpointID) Handle(params PutEndpointIDParams) (resp middleware.Resp
 	}
-	epTemplate := params.Endpoint
+
+	if err := server.CheckAPIMutationPermissions(params); err != nil {
```
Did you also consider writing a generic middleware to do this check? Having to manually add these seems brittle and the code generation seems more complicated than should be needed for something like this.

joamaki replied:
https://goswagger.io/use/middleware.html

joestringer replied:
I had a bit of an out-of-band discussion with @joamaki around this, and we think there's an option here. I had previously avoided this because I didn't want to get too deep into URL parsing in order to determine which APIs should be made available or not. It should be sufficient to categorize the APIs into groups like `endpoint`, `ipam`, `policy` etc. and expose those groups. This will make it easier to integrate as middleware rather than sprinkling authorization checks throughout all of the handlers.

joestringer committed:
daemon: Add warning for disabled APIs

The newly added API access flags are very flexible in the APIs that they
allow to be administratively disabled, but there are some options that
are really required for any standard Cilium operations. Add some basic
sanity checking before applying the user configuration in these cases,
as it could lead to unexpected results such as inability to keep the
Cilium agent running or inability to deploy new endpoints.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
daemon: Add Cells for Cilium API specifications

This preparatory commit introduces a new Cell for each Swagger API
Specification in order to allow reuse earlier on during the agent
initialization / lifecycle. No functional changes intended.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
daemon: Add flags for API access allowlist

Add new flag(s) to the daemon which restrict the ability for API clients
to call certain API endpoints. The new option is an allowlist of
Pascalized API endpoints that may be allowed, or optionally a Prefix
followed by the '*' character in order to allow a wildcard of API
endpoints, for example "*" for all API endpoints, or "Get*" for all GET
endpoints.

Set the default to allow all ("*") API modifications so that there is no
change in behaviour for existing deployments.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -2000,6 +1997,25 @@ func (d *Daemon) instantiateAPI() *restapi.CiliumAPIAPI {
 	restAPI.BgpGetBgpPeersHandler = NewGetBGPHandler(d.bgpControlPlaneController)
 
+	msg := "Required API option %s is disabled. This may prevent Cilium from operating correctly"
+	hint := "Consider enabling this API in --enable-api-access"
```
nit: Wrong flag name.

joestringer committed:
health: Rename variable to avoid pkg conflict

pkg/api will be used in an upcoming commit, so rename the variable in
this function to avoid that conflict.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
daemon: Add cilium-health API restrictions flag

Similar to the recent commit to administratively enable/disable support for
agent API endpoints, extend this support to the cilium-health API
exposed by nodes as well.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
operator: Rename variable to avoid pkg conflict

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
operator: Add operator API restrictions flag

For now this is not particularly important as the operator only has a
/healthz API to get the health of the operator. However, if we ever
extend this API in future then it could become useful to consistently
apply API restrictions via the new flag.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
api: Add flags to configure API access from spec

Use the autogenerated API specifications in order to populate "allowed
APIs" flags in the api specification objects for each API. This code
will be hooked in for usage in subsequent commits.

Signed-off-by: Joe Stringer <joe@cilium.io>
